Add contest judging system

venix12 opened the pull request:
resolves #903 (except the altering metadata part, i don't think it's in scope of that PR)
part of #10207

<details>
  <summary>Preview screenshots</summary>
  
  Contest judging panel, visible only to assigned judges on a per-contest basis.
  ![contest judge](https://cdn.discordapp.com/attachments/441583642758742016/1149100626836459610/image.png)

  Button in contest results that takes you to the judging results.
  ![contest results](https://cdn.discordapp.com/attachments/441583642758742016/1149100085087584318/image.png)
  
  Contest judging results. 
  ![contest judge results](https://cdn.discordapp.com/attachments/441583642758742016/1149102316637671424/image.png)
  
</details>

it could probably be more pretty :eye:

peppy commented:
@venix12 if possible could you add a video or screenshots showing what this is adding?

venix12 commented:
I've put the screenshots in the description.

notbakaneko reviewed (changes requested):
conflicts

peppy commented:
@notbakaneko are you able to resolve conflicts in cases like this on their behalf? It's a bit weird to not review something from our end for a  month then fire back that there's conflicts 😅 . Seems like an easy resolution.

notbakaneko commented on app/Models/ContestJudgeCategoryVote.php:
```diff
@@ -0,0 +1,33 @@
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+
+/**
+ * @property ContestJudgeCategory $category
```
For reference, stuff that goes through relations are readonly on the actual property itself and should be `@property-read`

venix12 commented:
resolved the conflicts :eyes:

notbakaneko commented on resources/js/entrypoints/contest-judge.tsx:
```diff
@@ -0,0 +1,18 @@
+import ContestEntryStore from 'stores/contest-entry-store';
+import { parseJson } from 'utils/json';
+
+const store = new ContestEntryStore();
+store.updateWithJson(parseJson('json-entries'));
```
This should go in the register callback, top level module code run when the module is first loaded, not when the page is loaded.

notbakaneko reviewed (changes requested):
Some quick notes first:
- is judging limited to specific types of contests?
- link to judging page from somewhere would be useful
- more context on the judging page on what's being judged would be useful, now it looks like a bunch of titles referring to what?

venix12 commented:
> Some quick notes first:
> 
> * is judging limited to specific types of contests?

It was traditionally used for beatmap contests only, don't really think there are currently any other use cases (?). If anything things can be adjusted when there's a need I'd say (actually would boil down to just adding judging results button to other listing types probably).
> * link to judging page from somewhere would be useful

Right, was an oversight from my side, added.
> * more context on the judging page on what's being judged would be useful, now it looks like a bunch of titles referring to what?

I added a question-mark-icon-tooltip kind of thing along with description field on categories, it should suffice i think?

notbakaneko commented:
> I added a question-mark-icon-tooltip kind of thing along with description field on categories

I mean on the judging page itself you'd see a bunch of titles with sliders and text boxes with no context of the thing being judged

venix12 commented:
Ah well, it's using entry masked name for title now. Assumption was the entries would be pulled and anonymized at database level using some external script (basically what's done at the moment I think) and then sent to judges so the judges should know which title goes for which entry (the masked name being set for author in beatmap file).

notbakaneko commented:
@cl8n do you have any feedback for this?

cl8n commented:
yeah I've tested this one a bit and will post a full review here. I think I can get to it this weekend but if not I'll give an update regardless

notbakaneko commented on resources/js/components/basic-select-options.tsx:
```diff
@@ -32,6 +32,8 @@ export default class RankingSelectOptions extends React.PureComponent<Props> {
 
   private href(id: number | null) {
```
At this point the component should probably just accept a callback to get the `href`.
Also, `modifiers` passed to `SelectOptions` should no longer be `spotlight`.

venix12 replied:
well, multiplayer one is still being casted from blade via registerComponents thing so not sure how to go with callback thing considering that :thinking:

notbakaneko commented on resources/views/contests/voting.blade.php:
```diff
@@ -32,6 +38,12 @@
             {{ $noVoteReason }}
         </div>
+
+        @if ($contestMeta->isJudged() && priv_check('ContestJudge', $contestMeta)->can())
```
Do admins automatically become judges? 🤔

venix12 replied:
guess this certain check could be added to the ``alwaysCheck`` stuff :eyes:

notbakaneko reviewed (changes requested):
Quick one for the things that stood out while I go through the rest 👀

notbakaneko commented on database/migrations/2023_07_20_153416_create_contest_judge_category_votes_table.php:
```diff
@@ -0,0 +1,39 @@
+     */
+    public function up(): void
+    {
+        Schema::create('contest_judge_category_votes', function (Blueprint $table) {
```
I feel like these are scores or grades, not votes 🤔

notbakaneko reviewed (changes requested):
The pivot tables with unique keys don't really need `id`s and should probably use a composite key instead.

`_votes` should probably be called `_scores` instead, especially if the grading criteria is always numeric and the table with the `comments` should probably be `contest_judge_comments`.

venix12 commented:
Personally feels like ``votes`` make more sense (for table with comments) given its usage, as it's got ``scores`` belonging to it (``contest_judge_vote_id`` with hasMany on ``votes``), pretty much always used that way. ``comments`` as a base for ``scores`` feels kinda off? 🤨

...unless you want me to disattach the hasMany relationship there and rather have both comments and scores relate to ``entry`` and ``user`` on its own?

notbakaneko commented:
I'll get back to this after the finishing going through the session and oauth verification stuff 👀

notbakaneko commented on app/Http/Controllers/ContestEntriesController.php:
```diff
@@ -5,20 +5,142 @@
+                    throw new InvariantException(osu_trans('contest.judge.validation.missing_score'));
+                }
+
+                $currentScore = ContestJudgeScore::where('contest_judge_vote_id', $vote->getKey())
```
```php
$vote->scores()->firstOrNew([
    'contest_judge_vote_id' => $vote->getKey(),
    'contest_scoring_category_id' => $category->getKey(),
])->fill([...])->save();
```

notbakaneko commented on app/Http/Controllers/ContestEntriesController.php:
```diff
@@ -5,20 +5,142 @@
+            }
+
+            foreach ($entry->contest->scoringCategories as $category) {
+                $score = $scores
```
Can use `Ds\Set` and get the intersection/diff for missing ids.
Or, just default missing values to 0.

notbakaneko commented on app/Http/Controllers/ContestEntriesController.php:
```diff
@@ -5,20 +5,142 @@
+            ->with('judgeVotes.scores')
+            ->findOrFail($id);
+
+        $updatedEntryJson = json_item($updatedEntry, 'ContestEntry', ['current_user_judge_vote.scores']);
+
+        return $updatedEntryJson;
```
can inline the return

notbakaneko commented on app/Http/Controllers/ContestEntriesController.php:
```diff
@@ -5,20 +5,142 @@
+            }
+        });
+
+        $updatedEntry = ContestEntry::with('judgeVotes')
```
`$entry->refresh()->load('judgeVotes.scores')`

notbakaneko reviewed (changes requested):
still checking....

venix12 commented on app/Http/Controllers/ContestEntriesController.php:
```diff
@@ -5,20 +5,142 @@
+        $comment = $params['comment'];
+
+        DB::transaction(function () use ($scores, $comment, $entry) {
+            $vote = $entry->judgeVotes->where('user_id', auth()->user()->getKey())->first();
```
thought the ``auth()->user()`` was the preference looking all around the code everywhere :thinking:

notbakaneko replied:
apparently using the facade is better for performance, so we've been moving back to that 👀

venix12 commented on resources/js/contest-judge-results/header.tsx:
```diff
@@ -0,0 +1,79 @@
+    const { contest, entry } = this.props;
+
+    const score = entry.results?.votes;
+    const maxScore = (contest.max_judging_score ?? 0) * (entry.judge_votes?.length ?? 0);
```
well max_judging_score is sum of scoring_category values so technically yes 🤨

all judges should be required to vote on all entries for things to make sense though so it shouldn't differ per-entry of one contest, ended up moving this part to backend anyway since it probably makes way more sense.

notbakaneko commented on resources/js/contest-judge/entry.tsx:
```diff
@@ -0,0 +1,191 @@
+  constructor(props: Props) {
+    super(props);
+
+    this.comment = props.entry.current_user_judge_vote?.comment ?? '';
+    this.initialVote = toJS(props.entry.current_user_judge_vote);
+    this.scores = props.entry.current_user_judge_vote?.scores ?? [];
```
Looking at this again, I think the whole `current_user_judge_vote` can be abstracted out to its own class where the scores are in a `Record` or `Map` and the comment is non-nullable?

notbakaneko commented on resources/js/components/basic-select-options.tsx:
```diff
@@ -32,6 +32,8 @@ export default class RankingSelectOptions extends React.PureComponent<Props> {
   private href(id: number | null) {
     switch (this.props.type) {
+      case 'judge_results':
+        return route('contest-entries.judge-results', { contest_entry: id });
```
oh yeah, `contest_entry` needs to be a non-nullable value

venix12 replied:
how would you go about resolving that? ...honestly not really ceratin about what to do with it🤔

notbakaneko replied:
just set it to 0 or return an empty href for now

notbakaneko commented on resources/js/contest-judge/entry.tsx:
```diff
@@ -0,0 +1,176 @@
+      return false;
+    });
+
+    if (
+      !scoresHaveChanged &&
+      this.currentVote.comment === this.initialVote.comment
+    ) return true;
+
+    return false;
```
Also should check if all the scores are set (I'm assuming the comment is not required here?)

```suggestion
    return !(
      this.currentVote.scores.size === this.props.store.scoringCategories.length
        && (scoresHaveChanged || this.currentVote.comment !== this.initialVote.comment)
    );
```

venix12 replied:
the comment is not required yes

notbakaneko reviewed (commented):
Will do for now, I guess 👀 

(I also re-dated the migrations)
